docs/proposals/runtime-image.md

```diff
@@ -0,0 +1,122 @@
+---
+title: runtime-image-support
+authors:
+  - "@sbose78"
+  - "@otaviof"
+status: design
+---
+
+Runtime-Image Support
+---------------------
+
+# Summary
+
+This enhancement proposal describes the support of runtime-image in build-v2 operator. A runtime-image is a composition based on the image just created by a third-party builder, used as input for the runtime-image.
+
+# Motivation
+
+With the runtime-image feature in place, it will be employed for the creation of _lean-images_, which takes the ability of picking only certain parts of the original image, creating a runtime-image with the reduced data-set.
+
+Additionally, creating images with a different base-image, and other advanced use-cases are also supported by this feature.
+
+# Proposal
+
+The sections below are describing the changes necessary to introduce the runtime-image.
+
+## API Changes
+
+A new attribute will be added on Build resource, as the following example:
+
+
+```yml
+---
+apiVersion: build.dev/v1alpha1
+kind: Build
+# ...
+spec:
+  runtime:
+    base:
+      image: runtime-base-image:latest
+    workDir: /workspace/source
+    env:
+      JAVA_HOME: /path/to/java-home
+    labels:
+      custom-image-label: value
+    directories:
+      - /path/to/directory
+      - /path/to/another/directory:/target/location
+    entrypoint:
+      - java
+      - -jar
+      - ...
+```
+
+The new attributes added are:
+
+- `spec.runtime.base`: specifies the runtime base-image to be used, using [Image](https://github.com/redhat-developer/build/blob/97012ab56417ce1691a70896d90e490ea6a4d23c/pkg/apis/build/v1alpha1/build_types.go#L58) as type;
+- `spec.runtime.workDir`: path to `WORKDIR` in runtime-image;
+- `spec.runtime.env`: runtime-image additional environment variables, key-value;
+- `spec.runtime.labels`: runtime-image additional labels, key-value;
+- `spec.runtime.directories`: list of directories paths to be copied to runtime-image, those can be defined as `<source>:<destination>` split by colon (`:`);
+- `spec.runtime.entrypoint`: entrypoint command, specified as a list;
+
```

Comment:
I think it would be of value to allow specifying a user. Rational is that many containers run as root by default while this is not allowed in all container runtime environments. Probably the generated Dockerfile should then create the user, `chown` all the mapped directories to the user and define the corresponding `USER` command.

Reply:
Should we default to USER 10001?

Reply:
Indeed, the ability of specifying the user is important to take care.

Given the `spec.runtime.base` can be any container-image, we need to create a flexible construction where the final user can specify configuration that will meet base-image's needs. So, jumping to how it would look like in `Build`, please consider:

```yml
---
apiVersion: build.dev/v1alpha1
kind: Build
# ...
spec:
  # ...
  runtime:
    # ...
    directories:
      - /workspace/source:/home/node/app
    # runtime-image user (USER) specification
    user:
      name: user
      # numerical UID to become part of `USER <UID>` in Dockerfile.runtime
      UID: 10001
      # triggers the chance of ownership in data copied to runtime-image
      chown: true
```

In this snippet, we would be able to create a Dockerfile with `USER` definition with `.spec.runtime.user.name`, while executing `chown` using the `UID`. The `chown` action could be disabled via the `chown` toggle.

However, there are implications with this approach, since the runtime base-image may not have `chown` present, and thus that command can fail.

### Exploring `RUN`

Having that said, in [classic `BuildConfig` we have a `postCommit` action](https://docs.openshift.com/container-platform/4.4/builds/understanding-buildconfigs.html), carrying out an arbitrary shell command. Should we also think about letting users conduct arbitrary shell commands as well? With a more flexible approach we may achieve the same outcome, but also allowing more use-cases to take place.

By exploring the `RUN` in `Dockerfile.runtime`, we could explore constructions like the following `Build` example:

```yml
---
apiVersion: build.dev/v1alpha1
kind: Build
# ...
spec:
  # ...
  runtime:
    # ...
    directories:
      - /workspace/source:/home/node/app
    # user employed on USER directive
    user: name
   # arbitrary commands to run after loading data
    run:
      - chown -R 10001:10001 /home/node/app
```

Please note the `.spec.runtime.user` can benefit of a simpler construction, while the chance of ownership is conducted in `.spec.runtime.run` section.

Reply:
@otaviof some ideas and open questions (not confident on what would be correct):

- do we want to allow to specify both user id and name ? Afaik, the id would be enough. But not sure if there are any circumstances that require a user with that user id to exist. If we allow to specify the user name (maybe as optional), would we also have another flag like `create: true|false` and then run `adduser` and `addgroup` commands?
- should we assume userId = groupId or allow specifying the groupId as well ?
- to change the ownership of the things copied through `directories`, we should do `COPY --chown userId:groupId source target` directly, then we do not need `RUN chown`. `kaniko` then calls golang's `os.chown()` (see https://github.com/GoogleContainerTools/kaniko/blob/cdbd8af0578c56e2801b57461e9f417f9479d303/pkg/util/fs_util.go#L750-L757) and I assume this does not need the `chown` command line.
- On the `RUN` approach. Difficulty there is what to do first? The `RUN` that creates the user referenced by `user: name` or the `USER` as the build originator expects its run steps to be performed by this user?

Reply:
> @otaviof some ideas and open questions (not confident on what would be correct):

Thanks for the inputs, I'm also thinking out loud here :-)

>     * do we want to allow to specify both user id and name ? Afaik, the id would be enough. But not sure if there are any circumstances that require a user with that user id to exist. If we allow to specify the user name (maybe as optional), would we also have another flag like `create: true|false` and then run `adduser` and `addgroup` commands?

Given the images may be completely different from each other we, unfortunately, cannot assume they will contain `adduser`/`addgroup` commands, or any other command. My suggestion here is to refrain from the responsibility creating the user, this action would be better fitted in the `RUN` suggestion.

A alternative would be to require the runtime image to already contain the UID/GID present.
 
>     * should we assume userId = groupId or allow specifying the groupId as well ?

We should amend the `Build` resource in a way we have GID informed. This is a important piece of information, which as not present in my first suggestion. 👍 

>     * to change the ownership of the things copied through `directories`, we should do `COPY --chown userId:groupId source target` directly, then we do not need `RUN chown`. `kaniko` then calls golang's `os.chown()` (see https://github.com/GoogleContainerTools/kaniko/blob/cdbd8af0578c56e2801b57461e9f417f9479d303/pkg/util/fs_util.go#L750-L757) and I assume this does not need the `chown` command line.

That's a good suggestion. Taking this example in consideration we could focus on having UID/GID informed, and use those on rendering the `Dockerfile.runtime` and using `--chown` flag.

>     * On the `RUN` approach. Difficulty there is what to do first? The `RUN` that creates the user referenced by `user: name` or the `USER` as the build originator expects its run steps to be performed by this user?

Okay, that's correct. Assuming the final user will be responsible by issuing `adduser`/`addgroup` commands, we may want to introduce two different run blocks. Before copying the `directories` and after.

Reply:
> Okay, that's correct. Assuming the final user will be responsible by issuing adduser/addgroup commands, we may want to introduce two different run blocks. Before copying the directories and after.

I do not think so. The final user can be assumed to be non-root and therefore cannot do `adduser`. Also, thinking about the use cases when I do `RUN` commands in Dockerfiles:

1. to install things
2. to download packages, extract them and make them executable
3. To add the non-root runtime user
(4. To compile the application, imo not applicable here as the build happened in a previous stage

All these things need to run as root. As most base images are using root, my proposal would be:

1. `RUN` happens before `USER`
2. Allow to specify a `user` with properties `user`, `group` and `chown` in the `runtime` section. The whole `user` section is optional. If it is defined, then `the nested user setting is mandatory. `group` is optional. `chown` should imo default to `true` (I have no good use case for it to be `false` honestly, if nobody has one, we could leave it out as well). The user can specify an id or name. The rational is that we can imo support both without effort for us. Named users and groups have to exist.

```yaml
runtime:
  directories:
      - /workspace/source:/home/node/app
  user:
    user: nonroot
    group: 1100
  run:
    - groupadd -g 1100 nonroot
    - useradd -u 1100 -g nonroot nonroot
```

would then become:

```Dockerfile

RUN groupadd -g 1100 nonroot && useradd -u 1100 -g nonroot nonroot

COPY --chown=nonroot:1100 --from=stage /workspace/source:/home/node/app
USER nonroot:1100
```

Both `COPY --chown` as well as `USER` support to specify ids and named entities.

BTW: I do not like the `runtime.user.user` thing (with `user` being twice in the property path). If somebody has a good alternative ...

Reply:
@SaschaSchwarze0  based on your input, I've updated the enhancement-proposal (on the last commit). Please consider it again.

Reply:
Implementation and documentation have been updated to support `user` and `run` attributes.

Reply:
Nice. :-)